src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -704,7 +711,14 @@ internal void FinishOperationSyncSuccess(int bytesTransferred, SocketFlags flags
                     socketError = FinishOperationConnect();
                     if (socketError == SocketError.Success)
                     {
-                        if (NetEventSource.IsEnabled) NetEventSource.Connected(_currentSocket!, _currentSocket!.LocalEndPoint, _currentSocket.RemoteEndPoint);
+                        if (NetEventSource.IsEnabled)
+                        {
+                            try
+                            {
+                                NetEventSource.Connected(_currentSocket!, _currentSocket!.LocalEndPoint, _currentSocket.RemoteEndPoint);
```

Comment:
Why does the nullable operator (!) appear with `_currentSocket` except for the .RemoteEndPoint?

Reply:
If a method is declared as taking a non-nullable value and you pass in something that may be nullable, the compiler will warn.  That's why the first `!` exists.  However, currently the compiler doesn't trust that the called method will throw an exception in such a case (e.g. the argument may be non-nullable but the method may not actually use the argument), so it doesn't use that to change the nullable tracking information for the variable; in this case, that means even though we put `!` on `_currentSocket!` as part of the first argument, it still considers `_currentSocket` to be "maybe null".  So then for the second argument, it warns when trying to dereference `_currentSocket` to access `LocalEndPoint`.  At that point, though, the compiler knows that by the time we get to `_currentSocket.RemoteEndPoint`, `_currentSocket` is not null... if it were null, the previous argument would have null ref'd.

(Note that there are current discussions about relaxing this tracking logic for C# 9, at which point we'd be able to remove the second `!` as well.)

Reply:
Thank you for the detailed explanation. It was non-obvious for me. I wonder if a comment above that line would help other readers of the code?

Reply:
We could.  Unfortunately `!`s are spread throughout the codebase, and there are many places where these kinds of inconsistencies show up.  They're already cluttering the code; I'm a bit weary of adding yet further clutter explaining why they exist.  My hope is that we'll be able to remove a bunch of them as the compiler's tracking improves.